rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,201 @@
+
+  // merges the props we declare internally and what is passed in
+  // by a consumer
+  const mergedProps = mergeProps(
+    {
+      // This is essentially shorthand for the `root` slot
+      ref: resolvedRef,
+      as: 'div',
+      icon: { as: 'span' },
+    },
+    defaultProps,
+    resolveShorthandProps(props, sampleShorthandProps),
+  );
+
+  if (checkedValues || onCheckedValuesChange) {
+    mergedProps.hasCheckMark = true;
+  }
+
+  // Anything else the component needs to manage its own state and render
+  [someState, setSomeState] = React.useState();
+  const { contextValue } = React.useContext();
+  React.useEffect(...);
+
+  // make an 'uber' state
+  const state = { someState, contextValue, ...mergedProps}
+
+  return state;
+};
+```
+
+#### useSampleClasses/useSampleStyles.ts
+
+Hook that accepts state and applies classnames to props and any shorthand slot props to style the component and its slots
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+```typescript
+import { getSlots } from '@fluentui/react-utils'
+
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.otherSlot {...slotProps.otherSlot} />
+    </slots.root>
+  );
+};
+```
+
+The `state` stores all information on the slots/tags/JSX that should be rendered after processing by the relevant hook. The `getSlots` utility extracts known slot information such as tag names and props.
+
+See below for how `state` might contain useful props for rendering.
+
+#### Sample.types.ts
+The below will probably be present in every component, other utility types are fair game if they are necessary.
+
+```typescript
+import * as React from 'react';
+import { ComponentProps, ShorthandProps } from '@fluentui/react-utils';
+
+// For a component all HTML attributes should be allowed to maximize consistencty with DOM
+export interface SampleProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  /** Icon slot as a shorthand */
+  icon?: ShorthandProps;
+}
+
+// Component state extends props and also adds other internal state used in the lifecycle of the component
+// For example for a popup component tracking an `open` state as boolean
+export interface MenuItemState extends SampleProps {
+  open: boolean;
+  
+  otherState: object;
+  
+  otherState: number;
+}
+```
+
+NOTE: Currently `ComponentProp` inherits from a `GenericDictionary` that is essentially `Record<string, any>` which is essentially `any` :-(
+
+### Pros and Cons
+
+#### Pros
+
+Since there is a separation of state, render, styling into different hooks/utility functions, it can be very easy for consumers to create their own custom component variants without needing to approach the library team and accommodate their custom requirements.
+
+It can be very easy to unit test each separate concern without too much wireframing/mocking.
```

Comment:
**thought (out off topic):** hmm is it really an 'Pro" - I'd avoid unit testing as much as possible as it tends to be quite brittle -> rather write functional tests (if those pieces work together well)

Yes this is for separate RFC 😅

Reply:
I am a believer of the church of 100% test coverage unfortunately, but let's address this in another RFC/discussion